python/cudf/cudf/core/buffer/weakrefable_buffer.py

```diff
@@ -0,0 +1,324 @@
+
+    In [10]: a is b
+    Out[10]: True
+    In [11]: b is c
+    Out[11]: True
+
+    This will be problematic as we cannot determine what is the count
+    of weak-references:
+
+    In [12]: weakref.getweakrefcount(x)
+    Out[12]: 1
+
+    Notice, though we want ``weakref.getweakrefcount`` to return ``3``, it
+    returns ``1``. So we need to work-around this by using an empty/no-op
+    callback:
+
+    In [13]: def custom_weakref_callback(ref):
+        ...:     pass
+        ...:
+
+
+    In [14]: d = weakref.ref(x, custom_weakref_callback)
+    In [15]: e = weakref.ref(x, custom_weakref_callback)
+    In [16]: f = weakref.ref(x, custom_weakref_callback)
+
+    Now there is an each unique weak-reference created:
+
+    In [17]: d
+    Out[17]: <weakref at 0x7f5beb03e360; to 'BufferWeakref' at 0x7f5c99ecd850>
+    In [18]: e
+    Out[18]: <weakref at 0x7f5bd15e3810; to 'BufferWeakref' at 0x7f5c99ecd850>
+    In [19]: f
+    Out[19]: <weakref at 0x7f5bd15f1f40; to 'BufferWeakref' at 0x7f5c99ecd850>
+
+    Now calling ``weakref.getweakrefcount`` will result in ``4``, which is correct:
+
+    In [20]: weakref.getweakrefcount(x)
+    Out[20]: 4
+
+    In [21]: d is not e
+    Out[21]: True
+
+    In [22]: d is not f
+    Out[22]: True
+
+    In [23]: e is not f
+    Out[23]: True
+    """  # noqa: E501
+    pass
+
+
+class RefCountableBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    _weak_ref: object
+    _proxy_ref: None | BufferWeakref
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    @classmethod
+    def _from_device_memory(cls: Type[T], data: Any) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._weak_ref = None
+        ret._proxy_ref = None
+        ret._zero_copied = False
+        ret._update_ref()
+        return ret
+
+    def _is_cai_zero_copied(self):
+        """
+        Returns a flag, that indicates if the Buffer has been zero-copied.
+        """
+        return self._zero_copied
+
+    def _update_ref(self):
+        """
+        Generate the new proxy reference.
+        """
+        # TODO: See if this can be merged into spill-lock
+        # once spilling and copy on write are merged.
+        self._proxy_ref = BufferWeakref(self._ptr, self._size)
+
+    def get_ref(self):
+        """
+        Returns the proxy reference.
+        """
+        if self._proxy_ref is None:
```

Comment:
When is this ever true? Shouldn't every `RefCountableBuffer` be initialized (constructed) with a `_proxy_ref`?